ldms/src/ldmsd/ldmsd.c

```diff
@@ -262,9 +301,45 @@ void __ldmsd_log(enum ldmsd_loglevel level, const char *fmt, va_list ap)
 		fprintf(log_fp, "%-10s: ", ldmsd_loglevel_names[level]);
 	}
 
-	vfprintf(log_fp, fmt, ap);
+	fprintf(log_fp, "%s", msg);
 	fflush(log_fp);
-	pthread_mutex_unlock(&log_lock);
+	return 0;
+}
+
+int log_actor(ev_worker_t src, ev_worker_t dst, ev_status_t status, ev_t ev)
+{
+	enum ldmsd_loglevel level = EV_DATA(ev, struct log_data)->level;
+	char *msg = EV_DATA(ev, struct log_data)->msg;
+	uint8_t is_rotate = EV_DATA(ev, struct log_data)->is_rotate;
+	int rc;
+
+	if (is_rotate)
+		rc = __logrotate();
+	else
+		rc = __log(level, msg);
+
+	free(msg);
+	ev_put(ev);
+	return rc;
+}
+
+void __ldmsd_log(enum ldmsd_loglevel level, const char *fmt, va_list ap)
+{
+	ev_t log_ev;
+	char *msg;
+	int rc;
+
+	log_ev = ev_new(log_type);
+	if (!log_ev)
+		return;
+	rc = vasprintf(&msg, fmt, ap);
```

Comment:
Perhaps a queue of 1024B message buffers that we can vsprintf into? This will reduce the overhead by avoiding the malloc/free.